validate-data.js

```diff
@@ -0,0 +1,212 @@
+    // The current obj being processed e.g. the obj with version 5.4
+    const currObj = data.data[i];
+    // The prev obj, e.g. 5.3
+    const prevObj = i > 0 ? data.data[i - 1] : null;
+    // The next obj, e.g. 6.0
+    const nextObj = i < (data.data.length - 1) ? data.data[i + 1] : null;
+    // The initial version 2 majors ahead e.g. for 4.13 this will be 6.0, for 5.2 this will be 7.0
+    const dmajObj = getVersionObj((getMajor(currObj) + 2) + '.0');
+    const isFinalMinor = nextObj && getMajor(currObj) !== getMajor(nextObj);
+    // Note that this is first minor for a major in data.json, not necessarily the actual first version
+    // For instance 5.2 might be the first minor in data.json, though 5.0 was the actual first version
+    // The !isFinalMinor check is for something like 4.13 which is the only version for CMS 4
+    const isFirstMinor = !isFinalMinor && (!prevObj || getMajor(prevObj) !== getMajor(currObj));
+    const isIntermediateMinor = !isFirstMinor && !isFinalMinor;
+
+    // variable which is reused for mutliple assertions
+    let expected;
+
+    // Update globally scope objected
+    version = currObj.version;
+
+    // Validate releaseDate - applies to all scenarios
+    if (hasBeenReleased(currObj)) {
+        scope = 'Released version';
+        if (currObj.releaseDateExtra !== null) {
+            error(`releaseDateExtra must be null`);
+        }
+    } else {
+        scope = 'Unlreleased version';
+        if (!currObj.releaseDate.match(/^(\d{4}-\d{2})$/)) {
+            error(`releaseDate must be in YYYY-MM format`);
+        }
+    }
+
+    // The following has some duplicated logic and extra nesting due to prioritising
+    // readabilitiy over conciseness
+    if (isFirstMinor) {
+        // e.g. 6.0, 7.0
+        scope = 'First minor version';
+        if (currObj.supportLength !== 'Standard') {
+            error(`supportLength must be "Standard"`);
+        }
+        // Standard + Full support - Approximately 6 months, until the next minor release is performed
+        if (nextObj) {
+            // Here we use "until the next minor release is performed"
+            if (currObj.partialSupport !== nextObj.releaseDate) {
+                error(`partialSupport must be "${nextObj.releaseDate}"`);
+            }
+        } else {
+            // Here we use "approximately 6 months", because there is no known next minor release date
+            expected = addTime(currObj.releaseDate, 6, 'months', false);
+            if (currObj.partialSupport !== expected) {
+                error(`partialSupport must be "${expected}"`);
+            }
+        }
+        if (!hasBeenReleased(currObj)) {
+            // This will be an unreleased next major, it needs an April-June release date
+            scope = 'Unreleased first minor version';
+            if (currObj.releaseDateExtra === null) {
+                error(`releaseDateExtra must not be null`);
+            }
+        }
```

Comment:
This is the same as the `partialSupport must be "${expected}"` scenario above - probably these should be combined so it's clearer they're both the same scenario.

Reply:
I'd rather have duplicate code than have one or two of these blocks as private methods, but not all the other blocks